project/Checkstyle.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import com.etsy.sbt.checkstyle.CheckstylePlugin.autoImport._
+import sbt._
+import sbt.Keys._
+
+object Checkstyle {
+  /*
+   ****************************
+   * Java checkstyle settings *
+   ****************************
+   */
+
+  private lazy val compileJavastyle = taskKey[Unit]("compileJavastyle")
+  private lazy val testJavastyle = taskKey[Unit]("testJavastyle")
+
+  def javaCheckstyleSettings(checkstyleFile: String): Def.SettingsDefinition = {
+    // Can be run explicitly via: build/sbt $module/checkstyle
+    // Will automatically be run during compilation (e.g. build/sbt compile)
+    // and during tests (e.g. build/sbt test)
+    Seq(
+      checkstyleConfigLocation := CheckstyleConfigLocation.File(checkstyleFile),
+      // if we keep the Error severity, `build/sbt` will throw an error and immediately stop at
+      // the `checkstyle` phase (if error) -> never execute the `check-report` phase of
+      // `checkstyle-report.xml` and `checkstyle-test-report.xml`. We need to ignore and throw
+      // error if exists when checking *report.xml.
+      checkstyleSeverityLevel := CheckstyleSeverityLevel.Ignore,
+
+      compileJavastyle := {
+        (Compile / checkstyle).value
+        javaCheckstyle(streams.value.log, checkstyleOutputFile.value)
+      },
+      (Compile / compile) := ((Compile / compile) dependsOn compileJavastyle).value,
+
+      testJavastyle := {
+        (Test / checkstyle).value
+        javaCheckstyle(streams.value.log, (Compile / target).value / "checkstyle-test-report.xml")
+      },
+      (Test / test) := ((Test / test) dependsOn (Test / testJavastyle)).value
+    )
+  }
+
+  private def javaCheckstyle(log: Logger, reportFile: File): Unit = {
```

Comment:
docs on each function explaining what it does. 
sbt scala code is hard to read and understnd .. so docs is important.